tests/keras/backend/reference_operations.py

```diff
@@ -526,6 +528,101 @@ def resize_volumes(x, depth_factor, height_factor, width_factor, data_format):
     return x
 
 
+def ndim(x):
+    dims = x.shape
+    if dims is not None:
+        return len(dims)
+    return None
+
+
+def batch_dot(x, y, axes=None):
+    if isinstance(axes, int):
+        axes = (axes, axes)
+    x_ndim = ndim(x)
+    y_ndim = ndim(y)
+    if axes is None:
+        # behaves like tf.batch_matmul as default
+        axes = [x_ndim - 1, y_ndim - 2]
+    if py_any([isinstance(a, (list, tuple)) for a in axes]):
+        raise ValueError('Multiple target dimensions are not supported. ' +
+                         'Expected: None, int, (int, int), ' +
+                         'Provided: ' + str(axes))
+    if x_ndim > y_ndim:
+        diff = x_ndim - y_ndim
+        y = np.reshape(y, np.concatenate([np.shape(y), [1] * (diff)], axis=0))
+    else:
+        diff = 0
+
+    if ndim(x) == 2 and ndim(y) == 2:
+        if axes[0] == axes[1]:
+            out = np.sum(np.multiply(x, y), axes[0])
+        else:
+            out = np.sum(np.multiply(np.transpose(x, [1, 0]), y), axes[1])
+    else:
+        out = np.tensordot(x, y, axes=axes)
+        for axis in [axes[0]]:
+            axis_list = np.arange(len(out.shape) - 1).tolist()
+            axis_list.insert(0, axis_list.pop(axis))
+            out = np.transpose(np.diagonal(out, axis1=0, axis2=axis),
+                               tuple(axis_list))
+    if diff:
+        if x_ndim > y_ndim:
+            idx = x_ndim + y_ndim - 3
+        else:
+            idx = x_ndim - 1
+        out = np.squeeze(out, tuple(range(idx, idx + diff)))
+    if ndim(out) == 1:
+        out = expand_dims(out, 1)
+    return out
+
+
+def int_shape(x):
+    return x.shape
+
+
+def cast(x, dtype):
+    return x.astype(dtype)
+
+
+def zeros(shape, dtype=None, name=None):
+    if dtype is None:
+        dtype = floatx()
+    v = np.zeros(shape=shape, dtype=dtype)
+    if py_all(list(v.shape)):
+        return variable(v, dtype=dtype, name=name)
```

Comment:
Is this condition necessary?

Reply:
I tried to stick as closely to the tf backend implementation as possible to avoid unexpected bugs.

Reply:
If you prefect simplicity I can remove the if py_all ....

Reply:
We don't need to follow tensorflow's code. The most important things are simplicity, readability and correctness. The tests are here for the correctness.